Always format to internal String in FmtPrinter

Mark-Simulacrum opened the pull request:
This avoids monomorphizing for different parameters, decreasing generic code
instantiated downstream from rustc_middle -- locally seeing 7% unoptimized LLVM IR
line wins on rustc_borrowck, for example.

We likely can't/shouldn't get rid of the Result-ness on most functions, though some
further cleanup avoiding fmt::Error where we now know it won't occur may be possible,
though somewhat painful -- fmt::Write is a pretty annoying API to work with in practice
when you're trying to use it infallibly.

rust-highfive commented:
Some changes occured to the CTFE / Miri engine

cc @rust-lang/miri

rust-highfive commented:
r? @oli-obk

(rust-highfive has picked a reviewer for you, use r? to override)

Mark-Simulacrum commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 81611d8cc69501c34dfeaa7df7e95469fde8b955 with merge e19edf03ef30826c4928f4a631b37e4871e43ab4...
<!-- homu: {"type":"TryBuildStarted","head_sha":"81611d8cc69501c34dfeaa7df7e95469fde8b955","merge_sha":"e19edf03ef30826c4928f4a631b37e4871e43ab4"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5254578217?check_suite_focus=true)
Build commit: e19edf03ef30826c4928f4a631b37e4871e43ab4 (`e19edf03ef30826c4928f4a631b37e4871e43ab4`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5254578217?check_suite_focus=true"},"merge_sha":"e19edf03ef30826c4928f4a631b37e4871e43ab4"} -->

rust-timer commented:
Finished benchmarking commit (e19edf03ef30826c4928f4a631b37e4871e43ab4): [comparison url](https://perf.rust-lang.org/compare.html?start=b8c56fa8c30821129b0960180f528d4a1a4f9316&end=e19edf03ef30826c4928f4a631b37e4871e43ab4).

**Summary**: This benchmark run shows 32 relevant regressions 😿 to instruction counts.
- Average relevant regression: 1.2%
- Largest regression in instruction counts: 3.0% on `full` builds of `deeply-nested-async check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

oli-obk commented:
While we reduce bootstrap time by 9s, these regressions seem fairly bad. I see three other solutions that could be tried:

* use dynamic dispatch
* explicitly monomorphize the two cases and expose only those, preventing duplication in every crate
* replace the generic param with `fmt::Formatter` and use some formatting wrapper for the string cases

Mark-Simulacrum commented:
I plan to dig into the regressions - my guess is that they're fixable along one of those lines. Maybe also preallocating some space in the buffer since we expect to print something.

Mark-Simulacrum commented:
I am pretty confident that the regressions here seem to be inlining decisions slightly changing -- looking at cachegrind diffs, it doesn't seem like the diffs belong to formatting code (mostly in ty_relate and similar generic code). 

I'm going to try and see if I can confirm that more closely with some disassembly, and maybe see if there's a few optimizations that can be applied regardless (e.g., by making use of the new information that our destination buffer is infallible), but my initial sense is that these regressions are likely somewhat outside the direct control here. We're just moving a bunch of code to being codegen'd just the one time, which naturally means there's a little less information for LLVM to make use of in some places -- things like which flags were set on FmtPrinter, for example -- which leads to slightly worse performance.

oli-obk commented on compiler/rustc_middle/src/ty/print/pretty.rs:
```diff
@@ -1562,24 +1562,24 @@ pub struct FmtPrinterData<'a, 'tcx, F> {
-    type Target = FmtPrinterData<'a, 'tcx, F>;
+impl<'a, 'tcx> Deref for FmtPrinter<'a, 'tcx> {
+    type Target = FmtPrinterData<'a, 'tcx>;
     fn deref(&self) -> &Self::Target {
```
It may be enough to slap `#[inline]` on all the functions in this file that appear in cachegrind for the first time after this PR

Mark-Simulacrum replied:
None of these functions appear in cachegrind -- they're not the direct problem. For example on stm32f4-check-full, we see a diff like this:

```
245,045,449  PROGRAM TOTALS

--------------------------------------------------------------------------------
Ir             file:function
--------------------------------------------------------------------------------
  372,484,116  ???:<core::iter::adapters::map::Map<core::iter::adapters::enumerate::Enumerate<core::iter::adapters::zip::Zip<core::iter::adapters::copied::Copied<core::slice::iter::Iter<rustc_middle::ty::subst::>
 -326,700,323  ???:rustc_middle::ty::relate::relate_substs::<rustc_infer::infer::equate::Equate>
  123,186,012  ???:<rustc_middle::ty::subst::SubstFolder as rustc_middle::ty::fold::TypeFolder>::fold_ty
   86,186,243  ???:rustc_middle::ty::relate::super_relate_tys::<rustc_infer::infer::equate::Equate>
  -55,808,143  ???:<rustc_middle::ty::Ty as rustc_middle::ty::fold::TypeFoldable>::super_fold_with::<rustc_middle::ty::subst::SubstFolder>
   30,561,468  ???:<alloc::collections::btree::map::entry::OccupiedEntry<rustc_infer::infer::region_constraints::Constraint, rustc_infer::infer::SubregionOrigin>>::remove_entry
  -27,764,057  ???:<alloc::collections::btree::node::Handle<alloc::collections::btree::node::NodeRef<alloc::collections::btree::node::marker::Mut, rustc_infer::infer::region_constraints::Constraint, rustc_infer:>
   20,749,456  ???:<&mut alloc::vec::Vec<ena::unify::VarValue<rustc_type_ir::TyVid>> as core::convert::AsRef<[ena::unify::VarValue<rustc_type_ir::TyVid>]>>::as_ref
   18,135,210  ???:<rustc_middle::mir::Rvalue>::ty::<rustc_middle::mir::Body>
   15,490,475  ???:<rustc_middle::ty::fold::RegionFolder as rustc_middle::ty::fold::FallibleTypeFolder>::try_fold_ty
  -12,892,781  ???:<rustc_middle::ty::Ty as rustc_middle::ty::fold::TypeFoldable>::super_fold_with::<rustc_middle::ty::fold::RegionFolder>
   12,520,606  ???:<rustc_infer::infer::combine::CombineFields>::instantiate
   12,024,623  ???:<rustc_data_structures::stable_hasher::StableHasher>::finish::<u128>
  -11,945,001  ???:<core::iter::adapters::map::Map<std::collections::hash::map::Iter<rustc_span::def_id::LocalDefId, rustc_hir::hir_id::ItemLocalId>, rustc_data_structures::stable_hasher::stable_hash_reduce<rust>
  -10,922,924  ???:<rustc_middle::ty::subst::GenericArg as rustc_middle::ty::relate::Relate>::relate::<rustc_infer::infer::nll_relate::TypeRelating<rustc_borrowck::type_check::relate_tys::NllTypeRelatingDelegate>
    9,837,670  ???:<alloc::vec::Vec<rustc_type_ir::TyVid> as alloc::vec::spec_from_iter::SpecFromIter<rustc_type_ir::TyVid, core::iter::adapters::filter_map::FilterMap<core::ops::range::Range<usize>, <rustc_infe>
   -8,795,699  ???:<rustc_infer::infer::sub::Sub as rustc_middle::ty::relate::TypeRelation>::relate_with_variance::<rustc_middle::ty::Ty>
    8,712,904  ???:<&mut rustc_middle::ty::relate::relate_substs<rustc_infer::infer::nll_relate::TypeRelating<rustc_borrowck::type_check::relate_tys::NllTypeRelatingDelegate>>::{closure#0} as core::ops::function>
    8,332,971  ???:<rustc_infer::infer::equate::Equate as rustc_middle::ty::relate::TypeRelation>::relate_with_variance::<rustc_middle::ty::Ty>
    8,262,191  ???:<rustc_infer::infer::sub::Sub as rustc_middle::ty::relate::TypeRelation>::tys
   -8,099,008  ???:<rustc_infer::infer::InferCtxt>::unsolved_variables
```

oli-obk replied:
Ah lol, yea, inlining cost boundaries are fun. Yea, just slap the attribute on all previously generic methods, that should make callers big enough again to stop them from being inlined. Alternatively add `inline(never)` with a comment to functions that disappeared from cachegrind.

Mark-Simulacrum replied:
That's not practical -- almost *every single* method became non-generic after removing the parameter; that's the whole reason for the big win on compile times. IMO trying to second-guess LLVM here on inlining (in either direction) is not the right approach, we're not likely to find a magic bullet.

The diff above is illustrative of typical inlining decisions slightly changing, but trying to recover from that by adding attributes isn't likely to really help that much -- the changes are too scattered and ultimately I suspect arise from things that we can't really influence easily.

oli-obk replied:
I thought the gains are from not instantiating the generics in 5 different crates and generating LLVM IR for them and optimizing them?

Having them only get compiled once is a different aspect than removing the implicit `#[inline]` that generic function have

Mark-Simulacrum replied:
`#[inline]` on all of these methods will -- AFAIK -- force generation in every crate that uses them; it's equivalent to generics in that respect. It also adds an inline hint, but that's a separate effect and unlikely to matter as much given the large size of many of the functions here. We save a little of course on generating just one copy per crate instead of 2-3, but that's still a higher cost.

Mark-Simulacrum commented:
@bors try @rust-timer queue

FWIW, I tried adding some eprintln! to the FmtPrinter::new and FmtPrinter::into_buffer functions, and neither are used at all for deeply-nested-async check builds. I think that further confirms that this is not in and of itself likely causing the regressions, rather that *other* functions are being shuffled between CGUs or similar due to these functions being omitted.

I've added a commit to preallocate a 64-byte buffer which might help with some benchmarks and rebased, let's see if we still see the same impacts.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

bors commented:
:hourglass: Trying commit 87bbb05815e1b44c940809f9ab261411bf0068fc with merge c65d716943343dccd7e2f9a1ad2455e7ebf9ae49...
<!-- homu: {"type":"TryBuildStarted","head_sha":"87bbb05815e1b44c940809f9ab261411bf0068fc","merge_sha":"c65d716943343dccd7e2f9a1ad2455e7ebf9ae49"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5260965109?check_suite_focus=true)
Build commit: c65d716943343dccd7e2f9a1ad2455e7ebf9ae49 (`c65d716943343dccd7e2f9a1ad2455e7ebf9ae49`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5260965109?check_suite_focus=true"},"merge_sha":"c65d716943343dccd7e2f9a1ad2455e7ebf9ae49"} -->

rust-timer commented:
Queued c65d716943343dccd7e2f9a1ad2455e7ebf9ae49 with parent e08d5693609a659e45025b8ea4dbd9efa342fa68, future [comparison URL](https://perf.rust-lang.org/compare.html?start=e08d5693609a659e45025b8ea4dbd9efa342fa68&end=c65d716943343dccd7e2f9a1ad2455e7ebf9ae49).

Mark-Simulacrum commented:
Preliminary local results suggest the 64-byte preallocation is not likely to have a significant impact here, so probably doesn't matter that much.

My suggested approach I think is still to call these relatively minor regressions justified -- they appear to be not directly related to the changes in this PR based on investigatory work I've done locally (cachegrind diffs, the fact that the code modified doesn't execute at all in some of the regressed benchmarks); I am inclined to take the win on bootstrap time and leave the runtime performance aside for this case. The folding code may have a few orthogonal wins that will buy us back some of this regression, but those wins are never going to really relate to this PR in a direct fashion.

rust-timer commented:
Finished benchmarking commit (c65d716943343dccd7e2f9a1ad2455e7ebf9ae49): [comparison url](https://perf.rust-lang.org/compare.html?start=e08d5693609a659e45025b8ea4dbd9efa342fa68&end=c65d716943343dccd7e2f9a1ad2455e7ebf9ae49).

**Summary**: This benchmark run shows 9 relevant regressions 😿 to instruction counts.
- Average relevant regression: 1.9%
- Largest regression in instruction counts: 2.8% on `incr-full` builds of `deeply-nested-async check`

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR led to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression

bors commented:
:pushpin: Commit 2ee6d55c62e9b62a20b5a87a7665e9836b879120 has been approved by `oli-obk`

<!-- @bors r=oli-obk 2ee6d55c62e9b62a20b5a87a7665e9836b879120 -->
<!-- homu: {"type":"Approved","sha":"2ee6d55c62e9b62a20b5a87a7665e9836b879120","approver":"oli-obk"} -->

bors commented:
:hourglass: Testing commit 2ee6d55c62e9b62a20b5a87a7665e9836b879120 with merge 8b9e41cd50112f711aef7f5d28d8458430bee486...
<!-- homu: {"type":"BuildStarted","head_sha":"2ee6d55c62e9b62a20b5a87a7665e9836b879120","merge_sha":"8b9e41cd50112f711aef7f5d28d8458430bee486"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5322006841?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/5322006841?check_suite_focus=true","builder_name":"checks-actions"} -->

Mark-Simulacrum commented:
@bors treeclosed=100 retry

Mark-Simulacrum commented:
@bors treeclosed-

bors commented:
:hourglass: Testing commit 2ee6d55c62e9b62a20b5a87a7665e9836b879120 with merge a26f75a3c88c9ff5b5bcc90d944e60a37ef8a0c3...
<!-- homu: {"type":"BuildStarted","head_sha":"2ee6d55c62e9b62a20b5a87a7665e9836b879120","merge_sha":"a26f75a3c88c9ff5b5bcc90d944e60a37ef8a0c3"} -->

rust-log-analyzer commented:
The job **`aarch64-gnu`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5321965427?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/8b9e41cd50112f711aef7f5d28d8458430bee486/checks/5321965427/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

Mark-Simulacrum commented:
@bors retry

bors commented:
:hourglass: Testing commit 2ee6d55c62e9b62a20b5a87a7665e9836b879120 with merge 4b043faba34ccc053a4d0110634c323f6c03765e...
<!-- homu: {"type":"BuildStarted","head_sha":"2ee6d55c62e9b62a20b5a87a7665e9836b879120","merge_sha":"4b043faba34ccc053a4d0110634c323f6c03765e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/5324283294?check_suite_focus=true)
Approved by: oli-obk
Pushing 4b043faba34ccc053a4d0110634c323f6c03765e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"oli-obk","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/5324283294?check_suite_focus=true"},"merge_sha":"4b043faba34ccc053a4d0110634c323f6c03765e"} -->

rust-timer commented:
Finished benchmarking commit (4b043faba34ccc053a4d0110634c323f6c03765e): [comparison url](https://perf.rust-lang.org/compare.html?start=3d127e2040b57157936f5f24e114a8b4c9a505ef&end=4b043faba34ccc053a4d0110634c323f6c03765e).

**Summary**: This benchmark run did not return any relevant results. 30 results were found to be statistically significant but too small to be relevant.

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression

rust-log-analyzer commented:
A job failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/5322556922?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/a26f75a3c88c9ff5b5bcc90d944e60a37ef8a0c3/checks/5322556922/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>
